In [ ]:
import pandas as pd

loans = pd.read_csv("LendingClubWithDescription.csv",low_memory=False)
loans

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0       1077501        NaN       5000         5000           4975.0   
1       1077430        NaN       2500         2500           2500.0   
2       1076863        NaN      10000        10000          10000.0   
3       1075358        NaN       3000         3000           3000.0   
4       1069639        NaN       7000         7000           7000.0   
...         ...        ...        ...          ...              ...   
126048    55716        NaN       3500         3500           3500.0   
126049   401654        NaN      12000        12000          12000.0   
126050   392759        NaN       4200         4200           4200.0   
126051   352554        NaN      21000        21000          21000.0   
126052   401243        NaN      25000        25000          25000.0   

              term int_rate  installment grade sub_grade  ...  \
0        36 months   10.65%       162.87     B        B2  ...   
1        60 months   15.27%        59.83     C        C4  ...   
2        36 months   13.49%       339.31     C        C1  ...   
3        60 months   12.69%        67.79     B        B5  ...   
4        60 months   15.96%       170.08     C        C5  ...   
...            ...      ...          ...   ...       ...  ...   
126048   36 months   24.99%       139.15     E        E4  ...   
126049   60 months   15.59%       289.21     C        C5  ...   
126050   36 months    7.99%       131.60     A        A4  ...   
126051   60 months   12.79%       475.57     C        C1  ...   
126052   36 months    8.99%       794.88     B        B1  ...   

       orig_projected_additional_accrued_interest  \
0                                             NaN   
1                                             NaN   
2                                             NaN   
3                                             NaN   
4                                             NaN   
...                                           ...   
126048                                        NaN   
126049                                        NaN   
126050                                        NaN   
126051                                        NaN   
126052                                        NaN   

       hardship_payoff_balance_amount hardship_last_payment_amount  \
0                                 NaN                          NaN   
1                                 NaN                          NaN   
2                                 NaN                          NaN   
3                                 NaN                          NaN   
4                                 NaN                          NaN   
...                               ...                          ...   
126048                            NaN                          NaN   
126049                            NaN                          NaN   
126050                            NaN                          NaN   
126051                            NaN                          NaN   
126052                            NaN                          NaN   

        debt_settlement_flag debt_settlement_flag_date settlement_status  \
0                          N                       NaN               NaN   
1                          N                       NaN               NaN   
2                          N                       NaN               NaN   
3                          N                       NaN               NaN   
4                          N                       NaN               NaN   
...                      ...                       ...               ...   
126048                     N                       NaN               NaN   
126049                     N                       NaN               NaN   
126050                     N                       NaN               NaN   
126051                     N                       NaN               NaN   
126052                     N                       NaN           

In [ ]:
labeled = pd.read_excel('Desc_labeled.xlsx')
labeled

id Self_label                                               desc
0   1216788      Other    Borrower added on 04/01/12 > This loan will ...
1   1856597      Other    Borrower added on 11/12/12 > I am in the pro...
2   8835950       Home    Borrower added on 11/11/13 > i had a home im...
3    995358       Home    Borrower added on 10/18/11 > Up front Cash i...
4   4564588      Other    Borrower added on 04/21/13 > Wheels and tire...
..      ...        ...                                                ...
95   351735      Other  I need money to buy something and even though ...
96  1540089   Business    Borrower added on 09/20/12 > Loan will be us...
97   753207      Other    Borrower added on 05/12/11 > How I plan to u...
98   508972      Other    Borrower added on 04/26/10 > My family is bu...
99  5967674       Home    Borrower added on 06/25/13 > For carpet  I a...

[100 rows x 3 columns]

In [ ]:
major = loans[loans["purpose"] == "major_purchase"].copy()
major

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
11      1069971        NaN       3600         3600           3600.0   
67      1067874        NaN       6000         6000           6000.0   
123     1067266        NaN       8400         8400           8150.0   
243     1064183        NaN       3000         3000           3000.0   
271      984879        NaN      18500        13250          13250.0   
...         ...        ...        ...          ...              ...   
125955  9867329        NaN      12000        12000          12000.0   
125978  9774870        NaN      32000        32000          31975.0   
125997   390449        NaN      14000        14000          14000.0   
126025   427092        NaN       8000         8000           8000.0   
126050   392759        NaN       4200         4200           4200.0   

              term int_rate  installment grade sub_grade  ...  \
11       36 months    6.03%       109.57     A        A1  ...   
67       60 months   12.69%       135.57     B        B5  ...   
123      60 months   16.77%       207.73     D        D2  ...   
243      36 months    7.51%        93.34     A        A3  ...   
271      60 months   11.71%       292.81     B        B3  ...   
...            ...      ...          ...   ...       ...  ...   
125955   60 months   12.85%       272.12     B        B4  ...   
125978   36 months    7.90%      1001.29     A        A4  ...   
125997   36 months   13.99%       478.42     C        C4  ...   
126025   36 months    6.99%       246.99     A        A3  ...   
126050   36 months    7.99%       131.60     A        A4  ...   

       orig_projected_additional_accrued_interest  \
11                                            NaN   
67                                            NaN   
123                                           NaN   
243                                           NaN   
271                                           NaN   
...                                           ...   
125955                                        NaN   
125978                                        NaN   
125997                                        NaN   
126025                                        NaN   
126050                                        NaN   

       hardship_payoff_balance_amount hardship_last_payment_amount  \
11                                NaN                          NaN   
67                                NaN                          NaN   
123                               NaN                          NaN   
243                               NaN                          NaN   
271                               NaN                          NaN   
...                               ...                          ...   
125955                            NaN                          NaN   
125978                            NaN                          NaN   
125997                            NaN                          NaN   
126025                            NaN                          NaN   
126050                            NaN                          NaN   

        debt_settlement_flag debt_settlement_flag_date settlement_status  \
11                         N                       NaN               NaN   
67                         N                       NaN               NaN   
123                        N                       NaN               NaN   
243                        N                       NaN               NaN   
271                        N                       NaN               NaN   
...                      ...                       ...               ...   
125955                     N                       NaN               NaN   
125978                     N                       NaN               NaN   
125997                     N                       NaN               NaN   
126025                     N                       NaN               NaN   
126050                     N                       NaN           

## 1)

In [ ]:
subset = major.sample(n=100, random_state=42)
subset.reset_index(drop=True, inplace=True)
subset

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0   1216788        NaN      13250        13250     13250.000000   36 months   
1   1856597        NaN      11000        11000     11000.000000   36 months   
2   8835950        NaN      25000        25000     25000.000000   36 months   
3    995358        NaN       4000         4000      4000.000000   36 months   
4   4564588        NaN       1125         1125      1125.000000   36 months   
..      ...        ...        ...          ...              ...         ...   
95   351735        NaN       5600         5600       737.826196   36 months   
96  1540089        NaN       5000         5000      5000.000000   36 months   
97   753207        NaN       7500         7500      7500.000000   60 months   
98   508972        NaN       8000         8000      8000.000000   36 months   
99  5967674        NaN       1000         1000      1000.000000   36 months   

   int_rate  installment grade sub_grade  ...  \
0    13.67%       450.74     B        B5  ...   
1     7.90%       344.20     A        A4  ...   
2    13.67%       850.44     B        B5  ...   
3    12.69%       134.18     B        B5  ...   
4    19.72%        41.65     D        D5  ...   
..      ...          ...   ...       ...  ...   
95   11.91%       185.76     C        C3  ...   
96   14.09%       171.11     B        B5  ...   
97   12.99%       170.61     C        C1  ...   
98   14.22%       274.28     C        C5  ...   
99    6.62%        30.71     A        A2  ...   

   orig_projected_additional_accrued_interest hardship_payoff_balance_amount  \
0                                         NaN                            NaN   
1                                         NaN                            NaN   
2                                         NaN                            NaN   
3                                         NaN                            NaN   
4                                         NaN                            NaN   
..                                        ...                            ...   
95                                        NaN                            NaN   
96                                        NaN                            NaN   
97                                        NaN                            NaN   
98                                        NaN                            NaN   
99                                        NaN                            NaN   

   hardship_last_payment_amount  debt_settlement_flag  \
0                           NaN                     N   
1                           NaN                     N   
2                           NaN                     N   
3                           NaN                     N   
4                           NaN                     N   
..                          ...                   ...   
95                          NaN                     N   
96                          NaN                     N   
97                          NaN                     N   
98                          NaN                     N   
99                          NaN                     N   

   debt_settlement_flag_date settlement_status settlement_date  \
0                        NaN               NaN             NaN   
1                        NaN               NaN             NaN   
2                        NaN               NaN             NaN   
3                        NaN               NaN             NaN   
4                        NaN               NaN             NaN   
..                       ...               ...             ...   
95                       NaN               NaN             NaN   
96                       NaN               NaN             NaN   
97                       NaN               NaN             NaN   
98                       NaN               NaN             NaN   
99                       NaN               NaN             NaN   

   settlement_amount settlement_

## Labeling a single loan description

First step: find a loan description that fits each of our 3  categories: house improvement, busines, and neither.

This will let us give specific examples for the prompt to work off of.

In [ ]:
#Home improvement
subset['desc'][19]

'  Borrower added on 12/11/12 > To purchase a Pellet fired Stove for our Homestead, to reduce the Electric heat cost that we pay on a monthly basis.  Estimated Savings during heat season would be approximately $1200. annually<br>'

In [ ]:
#Other
subset['desc'][6]

'  Borrower added on 04/08/10 > This loan is for the purchase of a used motorcycle.  I opted to use Lending Club as an alternative to financing the purchase due the high interest rates associated with traditional lenders.  I am very dependable when it comes to paying back anything.  I am the Chief Probation Officer for a rural county in northern California.  I have held the same job since I got out of the Marine Corps in 1994.<br/> Borrower added on 04/08/10 > In addition to my income, my wife is employed as an elementary school teacher for the last 7 years.  Both of our jobs are secure and combined we make approximately $152,000 per year.  I am as reliable as they come and look forward to hopefully being funded.<br/> Borrower added on 04/08/10 > I plan on using the funds to purchase a used motorcycle - my first. I  opted to go with Lending Club because the interest rates are much more reasonable than that of a traditional bank despite my excellent credit rating.  I am about as good as

In [ ]:
#Business 24?
subset['desc'][24]

'my business depends on good transportation and am currently in need of a new truck'

Let's start with a super-simple GenAI query before switching to follow best practices.

In [ ]:
test = subset['desc'][55]

In [ ]:
prelude = "Does the following loan description speak about a business related loan, home improvement or something else?\n\n"
query = prelude + test
print(query)

Does the following loan description speak about a business related loan, home improvement or something else?

  Borrower added on 10/12/11 > pleasure doing business with you<br/> Borrower added on 10/12/11 > I plan to use the money to start up a furniture business which me and my partner were planning for months. I just had to come up with a few dollars to put the order in to receive inventories we need to start. I have good credit, I have always made my payments on time and I am very serious about this business I'm about to begin. thank you<br/>


Great! Now let's use our sample code from the GenAISetup.ipynb notebook to run our query.

In [ ]:
key = ""

In [ ]:
from google import genai

client = genai.Client(api_key=key)

In [ ]:
response = client.models.generate_content(
    model='gemini-2.0-flash',
    contents=query
)
print(response.text)

The loan description clearly indicates that the loan is intended for **business purposes**. Specifically, the borrower states they plan to use the money to "start up a furniture business" and to order inventories needed to start the business.



## 2)

In [ ]:
prelude = """
You are a helpful lending club employee attempting to discern what type of loan is most prevelent on the site.

Your objective is to identify loans as one of the following:

Business related such as to start or continue business operations,

Home improvement related for a renevation, add on or other improvements,

Other which includes anything that does not fall into the first two categories.

You should respond with  "Business", "Home Improvement", or "Other" with no additional information provided.

Below you can find some examples:

Input: Borrower added on 12/11/12 > To purchase a Pellet fired Stove for our Homestead, to reduce the Electric heat cost that we pay on a monthly basis.  Estimated Savings during heat season would be approximately $1200. annually<br>
Output: Home Improvement
Input: my business depends on good transportation and am currently in need of a new truck
Output: Business
Input:  Borrower added on 04/08/10 > This loan is for the purchase of a used motorcycle.  I opted to use Lending Club as an alternative to financing the purchase due the high interest rates associated with traditional lenders.  I am very dependable when it comes to paying back anything.  I am the Chief Probation Officer for a rural county in northern California.  I have held the same job since I got out of the Marine Corps in 1994.<br/> Borrower added on 04/08/10 > In addition to my income, my wife is employed as an elementary school teacher for the last 7 years.  Both of our jobs are secure and combined we make approximately $152,000 per year.  I am as reliable as they come and look forward to hopefully being funded.<br/> Borrower added on 04/08/10 > I plan on using the funds to purchase a used motorcycle - my first. I  opted to go with Lending Club because the interest rates are much more reasonable than that of a traditional bank despite my excellent credit rating.  I am about as good as they come as it relates to loan repayments. I have always been very dependable/responsible with my money.  In addition to my income, my works full time as well as an elementary school teacher and combined we make nearly $150,000 per year.  Our jobs are very secure.  I am in law enforcement and am currrently the Chief of Probation in a rural county in northern California.  I have worked in the same department since I got out of the Marine Corps in 1994.  Thank you for taking the time to read this - I look forward to being funded.<br/>
Output: Other

Please respond for the following review, indicating "Business", "Home Improvement", or "Other" regarding the main reason for the loan request.

"""
query = prelude + test
print(query)


You are a helpful lending club employee attempting to discern what type of loan is most prevelent on the site.

Your objective is to identify loans as one of the following:

Business related such as to start or continue business operations,

Home improvement related for a renevation, add on or other improvements,

Other which includes anything that does not fall into the first two categories.

You should respond with  "Business", "Home Improvement", or "Other" with no additional information provided.

Below you can find some examples:

Input: Borrower added on 12/11/12 > To purchase a Pellet fired Stove for our Homestead, to reduce the Electric heat cost that we pay on a monthly basis.  Estimated Savings during heat season would be approximately $1200. annually<br>
Output: Home Improvement
Input: my business depends on good transportation and am currently in need of a new truck
Output: Business
Input:  Borrower added on 04/08/10 > This loan is for the purchase of a used motorcycle.  I

In [ ]:
response = client.models.generate_content(
    model='gemini-2.0-flash',
    contents=query
)
print(response.text)

Business



## Creating function to apply to larger amounts of data
* There was also some changes made to the 'prelude' due to errors found such empty strings returning full responses which took the form of loan descriptions not categories.

In [ ]:
def labelReview(desc):
    prelude = """
You are a helpful lending club employee attempting to discern what type of loan is most prevelent on the site.

Your objective is to identify loans as one of the following:

Business related such as to start or continue business operations,

Home improvement related for a renevation, add on or other improvements,

Other which includes anything that does not fall into the first two categories.

You should respond with  "Business", "Home", or "Other" with no additional information provided.

You should not misconstrude someone simply saying they are a good employee as meaning the loan is about business. The loan must be about improving or helping with business operations.

Below you can find some examples:

Input: Borrower added on 12/11/12 > To purchase a Pellet fired Stove for our Homestead, to reduce the Electric heat cost that we pay on a monthly basis.  Estimated Savings during heat season would be approximately $1200. annually<br>
Output: Home Improvement
Input: my business depends on good transportation and am currently in need of a new truck
Output: Business
Input:  Borrower added on 04/08/10 > This loan is for the purchase of a used motorcycle.
I opted to use Lending Club as an alternative to financing the purchase due the high interest rates associated with traditional lenders.
  I am very dependable when it comes to paying back anything.  I am the Chief Probation Officer for a rural county in northern
  California.  I have held the same job since I got out of the Marine Corps in 1994.<br/> Borrower added on 04/08/10 > In addition to my income,
  my wife is employed as an elementary school teacher for the last 7 years.  Both of our jobs are secure and combined we make approximately $152,000 per year.
  I am as reliable as they come and look forward to hopefully being funded.<br/> Borrower added on 04/08/10 > I plan on using the funds to purchase a used motorcycle - my first.
  I  opted to go with Lending Club because the interest rates are much more reasonable than that of a traditional bank despite my excellent credit rating.
  I am about as good as they come as it relates to loan repayments. I have always been very dependable/responsible with my money.
  In addition to my income, my works full time as well as an elementary school teacher and combined we make nearly $150,000 per year.  Our jobs are very secure.
  I am in law enforcement and am currrently the Chief of Probation in a rural county in northern California.  I have worked in the same department since I got out of the Marine Corps in 1994.
  Thank you for taking the time to read this - I look forward to being funded.<br/>
Output: Other

Please respond for the following review, indicating "Business", "Home", or "Other" regarding the main reason for the loan request.
These should be the only three responses.

If there is nothing below this line please respond with "Other" You should not give a response longer than 1 word.

"""
    query = prelude + desc
    response = client.models.generate_content(
        model='gemini-2.0-flash',
        contents=query
    )
    return response.text

labelReview(subset["desc"][0])

'Business\n'

In [ ]:
subset["desc"][0]

"  Borrower added on 04/01/12 > This loan will help me invest in a down payment for a truck that also invests in my future. When my credit report is pulled you'll see I've never been late on any monthly payments. My bills have always come first no matter my circumstances. I've been a loyal employee at same job for over 8 years.<br><br> Borrower added on 04/02/12 > My plan is to use the funds for a down payment on a semi truck. I'm a good borrower because I've never been late on any payments and my budget is that my bills always come first. I've been loyal to the same employer for over 8 years.<br>"

##3)
* The 'accuracy' of the llm model compared to the self-label is 87%(Actually 88% as one of the discrepencies is that the NA value for self-label was left empty compared to "other" for the LLM).
* This is a very acceptable accuracy, and gives good hope for the benefits of the process.

In [ ]:
to_label =subset[['id','desc']]

In [ ]:
#to_label.to_excel('Desc_to_label.xlsx', index=False)

In [ ]:
#labeled = pd.read_excel('Desc_labeled.xlsx')
#labeled

In [ ]:
from time import sleep
LLM_cat = []
i = 1
for description in subset["desc"]:
    LLM_cat.append(labelReview(description))
    print("working on desc "+str(i))
    i+=1
    sleep(6)
subset["LLM_cat"] = LLM_cat

working on desc 1
working on desc 2
working on desc 3
working on desc 4
working on desc 5
working on desc 6
working on desc 7
working on desc 8
working on desc 9
working on desc 10
working on desc 11
working on desc 12
working on desc 13
working on desc 14
working on desc 15
working on desc 16
working on desc 17
working on desc 18
working on desc 19
working on desc 20
working on desc 21
working on desc 22
working on desc 23
working on desc 24
working on desc 25
working on desc 26
working on desc 27
working on desc 28
working on desc 29
working on desc 30
working on desc 31
working on desc 32
working on desc 33
working on desc 34
working on desc 35
working on desc 36
working on desc 37
working on desc 38
working on desc 39
working on desc 40
working on desc 41
working on desc 42
working on desc 43
working on desc 44
working on desc 45
working on desc 46
working on desc 47
working on desc 48
working on desc 49
working on desc 50
working on desc 51
working on desc 52
working on desc 53
wo

##Issues with categorization
* Got one entry where the response was an entire desc in the style of a loan description. After further inspection this is the only entry in the subset that doesn't have a description. So we further specified the prompt to change this.

In [ ]:
#fixed
#subset.iloc[39]

In [ ]:
full = subset.merge(labeled, left_index=True, right_index=True)
full

id_x  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0   1216788        NaN      13250        13250     13250.000000   36 months   
1   1856597        NaN      11000        11000     11000.000000   36 months   
2   8835950        NaN      25000        25000     25000.000000   36 months   
3    995358        NaN       4000         4000      4000.000000   36 months   
4   4564588        NaN       1125         1125      1125.000000   36 months   
..      ...        ...        ...          ...              ...         ...   
95   351735        NaN       5600         5600       737.826196   36 months   
96  1540089        NaN       5000         5000      5000.000000   36 months   
97   753207        NaN       7500         7500      7500.000000   60 months   
98   508972        NaN       8000         8000      8000.000000   36 months   
99  5967674        NaN       1000         1000      1000.000000   36 months   

   int_rate  installment grade sub_grade  ... debt_settlement_flag_date  \
0    13.67%       450.74     B        B5  ...                       NaN   
1     7.90%       344.20     A        A4  ...                       NaN   
2    13.67%       850.44     B        B5  ...                       NaN   
3    12.69%       134.18     B        B5  ...                       NaN   
4    19.72%        41.65     D        D5  ...                       NaN   
..      ...          ...   ...       ...  ...                       ...   
95   11.91%       185.76     C        C3  ...                       NaN   
96   14.09%       171.11     B        B5  ...                       NaN   
97   12.99%       170.61     C        C1  ...                       NaN   
98   14.22%       274.28     C        C5  ...                       NaN   
99    6.62%        30.71     A        A2  ...                       NaN   

   settlement_status settlement_date  settlement_amount settlement_percentage  \
0                NaN             NaN                NaN                   NaN   
1                NaN             NaN                NaN                   NaN   
2                NaN             NaN                NaN                   NaN   
3                NaN             NaN                NaN                   NaN   
4                NaN             NaN                NaN                   NaN   
..               ...             ...                ...                   ...   
95               NaN             NaN                NaN                   NaN   
96               NaN             NaN                NaN                   NaN   
97               NaN             NaN                NaN                   NaN   
98               NaN             NaN                NaN                   NaN   
99               NaN             NaN                NaN                   NaN   

   settlement_term             LLM_cat     id_y Self_label  \
0              NaN          Business\n  1216788      Other   
1              NaN             Other\n  1856597      Other   
2              NaN  Home Improvement\n  8835950       Home   
3              NaN  Home Improvement\n   995358       Home   
4              NaN             Other\n  4564588      Other   
..             ...                 ...      ...        ...   
95             NaN             Other\n   351735      Other   
96             NaN          Business\n  1540089   Business   
97             NaN          Business\n   753207      Other   
98             NaN  Home Improvement\n   508972      Other   
99             NaN  Home Improvement\n  5967674       Home   

                                               desc_y  
0     Borrower added on 04/01/12 > This loan will ...  
1     Borrower added on 11/12/12 > I am in the pro...  
2     Borrower added on 11/11/13 > i had a home im...  
3     Borrower added on 10/18/11 > Up front Cash i...  
4     Borrower added on 04/21/13 > Wheels and tire...  
..                                                ...  
95  I need money to buy something and even thou

In [ ]:
#fixed
#weird = full[full['LLM_cat']=="Borrower added on 02/24/16 > We are good, loyal employees. My wife and I are looking to consolidate some debt and make some home improvements. We are purchasing a house and the previous owners were smokers. We would like to install hardwood floors, paint, and remove all the carpet. We are also looking to landscape the property, including building a fence for our dog.<br/>\nOutput: Home Improvement\n"]

In [ ]:
#fixed
#weird['desc_x']

In [ ]:
full['Self_label'].value_counts()

Self_label
Other       74
Home        17
Business     8
Name: count, dtype: int64

In [ ]:
full['LLM_cat'] = full['LLM_cat'].str.strip()
full['LLM_cat'] = full['LLM_cat'].str.replace(' Improvement', '')
full['LLM_cat'].value_counts()

LLM_cat
Other       68
Home        20
Business    12
Name: count, dtype: int64

In [ ]:
accuracy = sum(full['Self_label'] == full['LLM_cat'])/len(full)
print("Accuracy of LLM: "+str(accuracy*100)+"%")

Accuracy of LLM: 87.0%


In [ ]:
wrong = full[full['Self_label'] != full['LLM_cat']]
wrong = wrong[['id_x','desc_x','Self_label','LLM_cat']]
wrong = wrong.rename(columns={'id_x':'id','desc_x':'desc'})

## What categories were the self label and llm disagreeing on most?
* Seemingly the LLM was much quicker to discern a specific category compared to the self labeling. This could mean that it searches to put a loan into one of the two more defined categories which it has more information on.

In [ ]:
print(wrong['LLM_cat'].value_counts())
print(wrong['Self_label'].value_counts())

LLM_cat
Home        6
Business    4
Other       3
Name: count, dtype: int64
Self_label
Other    9
Home     3
Name: count, dtype: int64


In [ ]:
wrong.to_excel('Wrong.xlsx', index=False)

## Inconsistencies:
* Some where self-label was other but the LLM said Business it is ambiguous if the loan purchase is being used for work or just could be.
* When the self-label is other while the LLM said Home much of the time it is about buying a home which we interpreted as not counted in "Home Improvement", however this was too close for the LLM to discern.
* Those which the self-label was Home but the LLM said Other, were either specific ameneties(hot tub) that were not distinguished as for the home, or loans for equipment for home renevations which were possibly for one's own home or for a home renevation business.

## 4)
Testing Temperatures
0-2

In [ ]:
full = full[['id_x','desc_x','Self_label','LLM_cat']]
full = full.rename(columns={'id_x':'id','desc_x':'desc'})

In [ ]:
from google.genai import types


In [ ]:
def labelReview0(desc):
    prelude = """
You are a helpful lending club employee attempting to discern what type of loan is most prevelent on the site.

Your objective is to identify loans as one of the following:

Business related such as to start or continue business operations,

Home improvement related for a renevation, add on or other improvements,

Other which includes anything that does not fall into the first two categories.

You should respond with  "Business", "Home", or "Other" with no additional information provided.

You should not misconstrude someone simply saying they are a good employee as meaning the loan is about business. The loan must be about improving or helping with business operations.

Below you can find some examples:

Input: Borrower added on 12/11/12 > To purchase a Pellet fired Stove for our Homestead, to reduce the Electric heat cost that we pay on a monthly basis.  Estimated Savings during heat season would be approximately $1200. annually<br>
Output: Home Improvement
Input: my business depends on good transportation and am currently in need of a new truck
Output: Business
Input:  Borrower added on 04/08/10 > This loan is for the purchase of a used motorcycle.
I opted to use Lending Club as an alternative to financing the purchase due the high interest rates associated with traditional lenders.
  I am very dependable when it comes to paying back anything.  I am the Chief Probation Officer for a rural county in northern
  California.  I have held the same job since I got out of the Marine Corps in 1994.<br/> Borrower added on 04/08/10 > In addition to my income,
  my wife is employed as an elementary school teacher for the last 7 years.  Both of our jobs are secure and combined we make approximately $152,000 per year.
  I am as reliable as they come and look forward to hopefully being funded.<br/> Borrower added on 04/08/10 > I plan on using the funds to purchase a used motorcycle - my first.
  I  opted to go with Lending Club because the interest rates are much more reasonable than that of a traditional bank despite my excellent credit rating.
  I am about as good as they come as it relates to loan repayments. I have always been very dependable/responsible with my money.
  In addition to my income, my works full time as well as an elementary school teacher and combined we make nearly $150,000 per year.  Our jobs are very secure.
  I am in law enforcement and am currrently the Chief of Probation in a rural county in northern California.  I have worked in the same department since I got out of the Marine Corps in 1994.
  Thank you for taking the time to read this - I look forward to being funded.<br/>
Output: Other

Please respond for the following review, indicating "Business", "Home", or "Other" regarding the main reason for the loan request.
These should be the only three responses.

If there is nothing below this line please respond with "Other" You should not give a response longer than 1 word.

"""
    response = client.models.generate_content(
      model='gemini-2.0-flash',
      contents=prelude + desc,
      config=types.GenerateContentConfig(
      temperature= 0.0,
   )
)
    return response.text

In [ ]:
from time import sleep
LLM_cat0 = []
i = 1
for description in full["desc"]:
    LLM_cat0.append(labelReview0(description))
    print("working on desc "+str(i))
    i+=1
    sleep(6)
full["LLM_cat0"] = LLM_cat0

working on desc 1
working on desc 2
working on desc 3
working on desc 4
working on desc 5
working on desc 6
working on desc 7
working on desc 8
working on desc 9
working on desc 10
working on desc 11
working on desc 12
working on desc 13
working on desc 14
working on desc 15
working on desc 16
working on desc 17
working on desc 18
working on desc 19
working on desc 20
working on desc 21
working on desc 22
working on desc 23
working on desc 24
working on desc 25
working on desc 26
working on desc 27
working on desc 28
working on desc 29
working on desc 30
working on desc 31
working on desc 32
working on desc 33
working on desc 34
working on desc 35
working on desc 36
working on desc 37
working on desc 38
working on desc 39
working on desc 40
working on desc 41
working on desc 42
working on desc 43
working on desc 44
working on desc 45
working on desc 46
working on desc 47
working on desc 48
working on desc 49
working on desc 50
working on desc 51
working on desc 52
working on desc 53
wo

In [ ]:
full['LLM_cat0'] = full['LLM_cat0'].str.strip()
full

id                                               desc Self_label  \
0   1216788    Borrower added on 04/01/12 > This loan will ...      Other   
1   1856597    Borrower added on 11/12/12 > I am in the pro...      Other   
2   8835950    Borrower added on 11/11/13 > i had a home im...       Home   
3    995358    Borrower added on 10/18/11 > Up front Cash i...       Home   
4   4564588    Borrower added on 04/21/13 > Wheels and tire...      Other   
..      ...                                                ...        ...   
95   351735  I need money to buy something and even though ...      Other   
96  1540089    Borrower added on 09/20/12 > Loan will be us...   Business   
97   753207    Borrower added on 05/12/11 > How I plan to u...      Other   
98   508972    Borrower added on 04/26/10 > My family is bu...      Other   
99  5967674    Borrower added on 06/25/13 > For carpet  I a...       Home   

     LLM_cat  LLM_cat0  
0   Business  Business  
1      Other     Other  
2       Home      Home  
3       Home      Home  
4      Other     Other  
..       ...       ...  
95     Other     Other  
96  Business  Business  
97  Business  Business  
98      Home      Home  
99      Home      Home  

[100 rows x 5 columns]

In [ ]:
def labelReview2(desc):
    prelude = """
You are a helpful lending club employee attempting to discern what type of loan is most prevelent on the site.

Your objective is to identify loans as one of the following:

Business related such as to start or continue business operations,

Home improvement related for a renevation, add on or other improvements,

Other which includes anything that does not fall into the first two categories.

You should respond with  "Business", "Home", or "Other" with no additional information provided.

You should not misconstrude someone simply saying they are a good employee as meaning the loan is about business. The loan must be about improving or helping with business operations.

Below you can find some examples:

Input: Borrower added on 12/11/12 > To purchase a Pellet fired Stove for our Homestead, to reduce the Electric heat cost that we pay on a monthly basis.  Estimated Savings during heat season would be approximately $1200. annually<br>
Output: Home Improvement
Input: my business depends on good transportation and am currently in need of a new truck
Output: Business
Input:  Borrower added on 04/08/10 > This loan is for the purchase of a used motorcycle.
I opted to use Lending Club as an alternative to financing the purchase due the high interest rates associated with traditional lenders.
  I am very dependable when it comes to paying back anything.  I am the Chief Probation Officer for a rural county in northern
  California.  I have held the same job since I got out of the Marine Corps in 1994.<br/> Borrower added on 04/08/10 > In addition to my income,
  my wife is employed as an elementary school teacher for the last 7 years.  Both of our jobs are secure and combined we make approximately $152,000 per year.
  I am as reliable as they come and look forward to hopefully being funded.<br/> Borrower added on 04/08/10 > I plan on using the funds to purchase a used motorcycle - my first.
  I  opted to go with Lending Club because the interest rates are much more reasonable than that of a traditional bank despite my excellent credit rating.
  I am about as good as they come as it relates to loan repayments. I have always been very dependable/responsible with my money.
  In addition to my income, my works full time as well as an elementary school teacher and combined we make nearly $150,000 per year.  Our jobs are very secure.
  I am in law enforcement and am currrently the Chief of Probation in a rural county in northern California.  I have worked in the same department since I got out of the Marine Corps in 1994.
  Thank you for taking the time to read this - I look forward to being funded.<br/>
Output: Other

Please respond for the following review, indicating "Business", "Home", or "Other" regarding the main reason for the loan request.
These should be the only three responses.

If there is nothing below this line please respond with "Other" You should not give a response longer than 1 word.

"""
    response = client.models.generate_content(
      model='gemini-2.0-flash',
      contents=prelude + desc,
      config=types.GenerateContentConfig(
      temperature= 2.0,
   )
)
    return response.text

In [ ]:
from time import sleep
LLM_cat2 = []
i = 1
for description in full["desc"]:
    LLM_cat2.append(labelReview2(description))
    print("working on desc "+str(i))
    i+=1
    sleep(6)
full["LLM_cat2"] = LLM_cat2

working on desc 1
working on desc 2
working on desc 3
working on desc 4
working on desc 5
working on desc 6
working on desc 7
working on desc 8
working on desc 9
working on desc 10
working on desc 11
working on desc 12
working on desc 13
working on desc 14
working on desc 15
working on desc 16
working on desc 17
working on desc 18
working on desc 19
working on desc 20
working on desc 21
working on desc 22
working on desc 23
working on desc 24
working on desc 25
working on desc 26
working on desc 27
working on desc 28
working on desc 29
working on desc 30
working on desc 31
working on desc 32
working on desc 33
working on desc 34
working on desc 35
working on desc 36
working on desc 37
working on desc 38
working on desc 39
working on desc 40
working on desc 41
working on desc 42
working on desc 43
working on desc 44
working on desc 45
working on desc 46
working on desc 47
working on desc 48
working on desc 49
working on desc 50
working on desc 51
working on desc 52
working on desc 53
wo

In [ ]:
full['LLM_cat2'] = full['LLM_cat2'].str.strip()
full

id                                               desc Self_label  \
0   1216788    Borrower added on 04/01/12 > This loan will ...      Other   
1   1856597    Borrower added on 11/12/12 > I am in the pro...      Other   
2   8835950    Borrower added on 11/11/13 > i had a home im...       Home   
3    995358    Borrower added on 10/18/11 > Up front Cash i...       Home   
4   4564588    Borrower added on 04/21/13 > Wheels and tire...      Other   
..      ...                                                ...        ...   
95   351735  I need money to buy something and even though ...      Other   
96  1540089    Borrower added on 09/20/12 > Loan will be us...   Business   
97   753207    Borrower added on 05/12/11 > How I plan to u...      Other   
98   508972    Borrower added on 04/26/10 > My family is bu...      Other   
99  5967674    Borrower added on 06/25/13 > For carpet  I a...       Home   

     LLM_cat  LLM_cat0  LLM_cat2  
0   Business  Business  Business  
1      Other     Other     Other  
2       Home      Home      Home  
3       Home      Home      Home  
4      Other     Other     Other  
..       ...       ...       ...  
95     Other     Other     Other  
96  Business  Business  Business  
97  Business  Business  Business  
98      Home      Home      Home  
99      Home      Home      Home  

[100 rows x 6 columns]

In [ ]:
full['all_matching'] = full.apply(lambda x: x.LLM_cat == x.LLM_cat0 == x.LLM_cat2, axis = 1)

In [ ]:
sum(full['all_matching'])

98

## Changes with different temperatures
*  Across all 3 temperatures there is only 2 entries where the different calls gave different responses. This is very promising as it could indicate that the prompt is particular enough to give consistent results.

In [ ]:
.76 + 3.03+.19 + 7.56+ 4359.355

4370.8949999999995

In [ ]:
loans['desc_c_count']= loans['desc'].str.len()

In [ ]:
input_characters = loans['desc_c_count'].sum()
print("input characters: "+str(input_characters))
input_tokens = input_characters/4
print("input tokens: "+str(input_tokens))

input characters: 30249598
input tokens: 7562399.5


##5) Time and Cost Estimates
### Free Tier:
##### Cost:
* Luckily this part is easy it is free though this will lead to a much more time required to process this huge amount of data.
##### Time Required:
* At minimum this process would take 85 days. This daunting number comes from the fact that we would require a separate request for each of the 126,053 loan descriptions, and the free tier has a limit of 1,500 requests per day.

### Tier 1:
##### Cost:
* The cost of the paid tier would be roughly $4370.89. The bulk of this cost comes from the grounding used in Gemini and the cost per thousand requests after a certain limit. The calculation was done by dividing te number of tokens (roughly 7562400) dividing by 1000000 and multiplying across the costs per million tokens and per thousand requests. Calculation as follows:(input price).76 + (output price)3.03+(context caching).19 + (context caching storage)7.56+ (grounding)4359.355
##### Time Required:
* This is where the tier up comes with large perks. As there is no daily limit on requests in this tier, what we have to worry about is the requests per minute limit which is 2,000. This should allow the entire dataset to be completed in just under 1 hour and 4 minutes. The token per minute rate should not be an issue as there is an estimated total of under 8 million tokens for the entire dataset and the token per minute limit is 4 million.